### PR TITLE
[DesignPR] Controller Unit Testability Part 2

### PR DESCRIPTION
**Pull request by ianhong**
@rynowak  and @yishaigalatzer 

Could you take a look at this design PR? I'm also attaching the summary of this PR. 

Thanks

What are covered or tested in this Design PR
1. ViewBag
2. ModelState
3. Context.Request.GetFormAsync();formCollection.GetValues("PromoCode").FirstOrDefault()
7. Url.Action
5. Context.RequestAborted
9. Url.IsLocalUrl(returnUrl

Not yet implemented:
1. Context.User.Identity.GetUserName(), Context.User.Identity.GetUserId(), User.Identity.IsAuthenticated
2. Context.Request.Scheme
3. Response.SignOut
4. Need to expand the UrlAction callback signature
5. Context.Request.Cookies.Get("Session"); Context.Request.Cookies.Append("Session", cartId);
9. Context.Request.ReadFormAsync();formParameters["RequestVerificationToken"];

External Components to Test:
1. DbContext
2. IMemoryCache
3. IConnectionManager
4. UserManager<ApplicationUser>
5. SignInManager<ApplicationUser>
6. IHostingEnvironment
7. AntiForgery
8. IOptions
9. RoleManager<IdentityRole>
10. IFlashStrategy, ILogger

Exploratory Test
1 Adding controller unit tests for MusicStore
2. Updating unit tests for KExpense

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+				var formCollection = Context.Request.ReadFormAsync();
+				var viewName = formCollection.Result.GetValues(formKey).FirstOrDefault();
+
+				return View(viewName);
+			}
+
+			public IActionResult UsingUrlAction(string action)
+			{
+				return View(Url.Action(action));
+			}
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: Would really like to see some more real-world examples here. These are a bit simple and are contrived towards the features that you targeted. Can we take some code from music store and mock out the db to use that as a starting point?

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+				((MvcControllerTestHelperHttpContext)HttpContext).OnRequestAborted = callback.OnRquestAborted;
+				((MvcControllerTestHelperHttpContext)HttpContext).OnRequestFormCollection = callback.OnRequestFormCollection;
+				((MvcControllerTestHelpeUrlHelper)controller.Url).OnAction =
+						(action, controllerName, values, protocol, host, fragment) => callback.OnUrlAction(action);
+			}
+
+			return controller;
+		}
+
+		private class MvcControllerTestHelperHttpContext : HttpContext
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: We should use `DefaultHttpContext` here. My guess is that we want any testing helper code to go into a separate project and not `Mvc.Core`. Then it's no trouble to introduce the dependency on the package that `DefaultHttpContext` lives in.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Core;
+using Microsoft.AspNet.Http.Core.Collections;
+using Microsoft.AspNet.Http.Security;
+using Microsoft.AspNet.Routing;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+// Not sure if this is the right location/namespace for this..
+namespace Microsoft.AspNet.Mvc.Core
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: Discuss with @yishaigalatzer - my assumption is that we want another separate package - something like `Microsoft.AspNet.Mvc.Testing`. 

In general though there is no product code in the `Microsoft.AspNet.Mvc.Core` namespace. It should be `Microsoft.AspNet.Mvc` if it's something that a user will need to find easily, otherwise `Microsoft.AspNet.Mvc.<FeatureName>` if it's something obscure.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+{
+	public class MvcControllerUnitTestHelper
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: What's the value of this class being instance-based? I don't see any of your examples leveraging that.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+			{
+				return OnAction(action, controller, values, protocol, host, fragment);
+			}
+
+			public string Content(string contentPath)
+			{
+				throw new NotImplementedException();
+			}
+
+			public bool IsLocalUrl(string url) // Can it be a static method or the UrlHelper class or UrlHelperExtensions??
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: You can subclass `UrlHelper` (with some tweaks) - or what's probably better is to move this to a static method and have `UrlHelper` and the testable version both call that method.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelperCallback.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+		public Func<CancellationToken> OnRquestAborted
+		{
+			get;
+			set;
+		}
+		public Func<FormCollection> OnRequestFormCollection
+		{
+			get;
+			set;
+		}
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelperCallback.cs]: Take a look at https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Core/DefaultHttpContext.cs

In general all of the `HttpContext` stuff is get/set properties. In general setting a property is much simpler to get your head around than using callbacks, so we should leverage that there the default is testable. If you find things in that space that don't work well let's loop in the appropriate folks and improve it.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+using Microsoft.AspNet.Routing;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+// Not sure if this is the right location/namespace for this..
+namespace Microsoft.AspNet.Mvc.Core
+{
+	public class MvcControllerUnitTestHelper
+	{
+		public MvcControllerUnitTestHelper() : this(new ServiceCollection(), new RouteData())
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: Completely pointless, but the spacing here looks odd.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 3 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: Yes this should be in another package

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+		}
+
+		// Not sure if this HttpContext property will be usefull..
+		public HttpContext HttpContext
+		{
+			get;
+			set;
+		}
+
+		public Controller Initialize(Controller controller, MvcControllerUnitTestHelperCallback callback)
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: a generic method will be much more useful

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 4 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: the name can be just `ControllerTestHelper`

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 5 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: I think the latter (static method) should be better. I tried to subclass UrlHelper, but it quires users to handle router, routedata, routerbuild and so on. I think that will make this scenario much harder.

Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: Probably not  the best idea to subclass the `UrlHelper` subclassing is useless I hear.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 4 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: Agree. It was static before, but while I'm trying to subclass UrlHelper, I made it a regular type and forgot to reverse. Thanks.

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNet.Http.Core.Collections;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core.Test
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: This will move to a new test project as well

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs (hunk 1 above)
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: @rynowak In fact, my plan is adding controller unit tests for MusicStore once this feature is done. So, that's why these unit tests are simpler. I'm wondering if you still want to make these unit tests more realistic in addition to that??

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+			// Arrange
+			var controller = new TestabilityController();
+
+			var callback = new MvcControllerUnitTestHelperCallback()
+			{
+				OnRquestAborted = () => new CancellationToken(true)
+			};
+
+			// Act
+			new MvcControllerUnitTestHelper().Initialize(controller, callback);
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: this will just be:

```c#
ControllerTestHelper.Create<TestabilityController)(callback);
```

or - 

```c#
ControllerTestHelper.Create<TestabilityController)(new ControllerTestCallback() // note the name change
    {
        On RequestAborted = () => new CancellationToken(true),
    });
```

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+			var viewResult = controller.UsingContextRequestAborted();
+
+			// Assert
+			Assert.Equal("RequestAborted", viewResult.ViewName);
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: This is a bad example - if the Request is aborted, you will never return a view

File: test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs (hunk 1 above)
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/UnitTestPrototype.cs]: Its probably better to build a test sample on top of music store and send as a separate PR (in the Music Store repo). I agree though the scenarios should be more realistic.

File: src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs (hunk 8 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs]: @yishaigalatzer I think non-generic version can be more useful because many of controllers take parameters for IDbContxt, or ICacheManager and so on..
Do you think we should provide both of non-generic version and generic version?

**Comment by ianhong** [comment]
@rynowak and @yishaigalatzer 

I updated this PR per your comments. This is the summary:
1) Changed the IUrlHeper interface to remove IsLocalUrl and made it a static method inside UrlHelper. I think I should send a seperate PR for this. 
2) The --Callback name changed to --Context to accommodate the simple value properties.
3) Used DefaultHttpContext 
4) Helper is now a static class

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+				helperContext.HttpContext.Abort();
+			}
+
+			((MockUrlHelper)controller.Url).OnAction
+				= (action, controllerName, values, protocol, host, fragment)
+				=>
+				{
+					helperContext.RequestScheme = protocol;
+					helperContext.RequestHost = host;
+					helperContext.RequestFragment = fragment;
```
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: I updated helperContext here to simplify the OnUrlAction callback. not sure if this a good design or not. Otherwise, the signature of OnUrlAction will be Func<string, string, object, string, string, string, string> OnUrlAction. I think this is too long for users.

File: test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs [hunk 13]
```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: In this design PR, I didn't make these test more realistic. I'll do it in the actual PR.

File: test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+			{
+				RequestAborted = true
+			};
+
+			// Act
+			ControllerTestHelper.Initialize(controller, context);
+			var viewResult = controller.UsingContextRequestAborted();
+
+			// Assert
+			Assert.Null(viewResult);
```
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: What is the correct behavior when requests are aborted?? This Context.RequestAborted is used in many places in MusicStore like https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Controllers/CheckoutController.cs
I can't find any Http status code for it too.

File: src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs [hunk 15]
```diff
@@ -68,8 +68,13 @@ public UrlHelper(IScopedInstance<ActionContext> contextAccessor, IActionSelector
             return GenerateUrl(protocol, host, path, fragment);
         }
 
-        /// <inheritdoc />
-        public bool IsLocalUrl(string url)
+        /// <summary>
+        /// Returns a value that indicates whether the URL is local.
+        /// </summary>
+        /// <param name="url">The URL.</param>
+        /// <returns>true if the URL is local; otherwise, false.</returns>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs]: `<c>true</c>` same for false.

File: src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs [hunk 16]
```diff
@@ -68,8 +68,13 @@ public UrlHelper(IScopedInstance<ActionContext> contextAccessor, IActionSelector
...
+
+        public static bool IsLocalUrl(string url)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs]: @yishaigalatzer we were discussing making this method back into a static. It's a security feature and having it show up as mockable seemed bad.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+namespace Microsoft.AspNet.Mvc.Testing
+{
+	public static class ControllerTestHelper
+	{
+		public static Controller Initialize(Controller controller, ControllerTestHelperContext context)
+		{
+			return Initialize(controller, new ActionDescriptor(), context);
+			// Question: Can we build (fully or partially) ActionDescriptor() from tyep information??
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: To answer your question, we can't do this in any sort of accurate way.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+		private static DefaultHttpContext CreateHttpContext(IServiceProvider serviceProvider)
+		{
+			var featureCollection = new FeatureCollection();
+			var featureContext = new FeatureContext();
+
+			featureCollection.Add(typeof(IHttpRequestLifetimeFeature), new FeatureContext());
+			featureCollection.Add(typeof(IHttpRequestFeature), new FeatureContext());
+
+			var httpContext = new DefaultHttpContext(featureCollection)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: None of this should be needed. Just `new DefaultHttpContext()`

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+			{
+				RequestServices = serviceProvider
+			};
+
+			return httpContext;
+		}
+
+		private static Controller Initialize(Controller controller, ActionDescriptor actionDescriptor,
+			ControllerTestHelperContext helperContext)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: don't break the line in the middle of the parameter list. Either they all go on their own line, or all on the same.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+			var featureCollection = new FeatureCollection();
+			var featureContext = new FeatureContext();
+
+			featureCollection.Add(typeof(IHttpRequestLifetimeFeature), new FeatureContext());
+			featureCollection.Add(typeof(IHttpRequestFeature), new FeatureContext());
+
+			var httpContext = new DefaultHttpContext(featureCollection)
+			{
+				RequestServices = serviceProvider
+			};
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: We should probably set `ApplicationServices` also

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: My suggestion is to make `MockUrlHelper` a property on the `HelperContext` that gets initialized by default. Then you can add multiple overloads of things like `OnAction` to that class.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+			get;
+			set;
+		}
+
+		// Not sure if this HttpContext property will be usefull..
+		public HttpContext HttpContext
+		{
+			get;
+			set;
+		}
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: These are both absolutely useful. My suggestion for all of these is to pre-intialize them so the user can just set values directly.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+
+		public IServiceProvider ServiceProvider
+		{
+			get;
+			set;
+		}
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: Both of these are potentially useful - `ServiceCollection` to manually add services that will be resolved.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+
+		public FormCollection RequestFormCollection
+		{
+			get;
+			set;
+		}
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: If you pre-intialize `HttpContext`, then there's no real need for any of these, because you can just do it through the context.

`context.HttpContext.Request.Scheme = "https";`

File: src/Microsoft.AspNet.Mvc.Testing/FeatureContext.cs [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Core;
+using Microsoft.AspNet.Http.Interfaces;
+
+namespace Microsoft.AspNet.Mvc.Testing
+{
+	internal class FeatureContext :
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/FeatureContext.cs]: this shouldn't be needed

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+
+		public Func<string, string, object, string> OnUrlAction
+		{
+			get;
+			set;
+		}
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: suggest removing this and adding `MockUrlHelper` directly.

File: test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs (hunk 13 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: It's **really** hard to get a grasp on whether or not this design stands up without some compelling tests and controller code to look at.

Examples:
https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Controllers/AccountController.cs#L227
https://github.com/aspnet/MusicStore/blob/master/src/MusicStore.Spa/Apis/AlbumsApiController.cs#L144

File: test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+				=>
+				{
+					Assert.Equal("controller1", controllerName);
+					Assert.Equal(null, values);
+					Assert.Equal("http", helperContext.RequestScheme);
+					Assert.Equal("host1", helperContext.RequestHost);
+					Assert.Equal("fragment1", helperContext.RequestFragment);
+					Assert.Equal("redirect", action);
+					return action;
+				}
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: This isn't really right - why is the helperContext being validated in a test? You want to validate that route values passed to URL helper are valid, and then generate a fake URL so the action can proceed.

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs (hunk 18 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: The main reason of using this feature collection is providing users with a way to mock  Context.RequestAborted.IsCancellationRequested. I'll talk to @Tratcher if there is a better way to do this. In the meantime I'd like to keep this to unblock myself..

Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]:
```C#
var context = new DefaultHttpContext();
context.SetFeature<IHttpRequestLifetimeFeature>(myMockLifetimeFeature);
...
class MyMockLifetimeFeature : IHttpRequestLifetimeFeature
{
...
}
```

Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: @Tratcher Thanks,
That looks not much different from new DefaultHttpContext(myMockLifetimeFeature)..I think what @rynowak meant was is there any simpler way to do that like RequestAborted settable properties?

Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: If you try my sample it will be simpler than what you already have, you don't have to mock IHttpRequestFeature.

Making HttpContext.RequestAborted and/or IHttpRequestLifetime.RequestAborted settable could be done, but we'd have to work out what that would mean for IHttpRequestLifetimeFeature.Abort().

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs (hunk 12 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: @rynowak How about converting OnUrlAction property to the methods of HelperContext such as. In this way, we don't have to make MockUrlHelper public...?? 
class HelperContext {
               // 
		public void OnUrlAction(Func<string> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}

		public void OnUrlAction(Func<string, string> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}
		public void OnUrlAction(Func<string, string, object> actionCallBack)
		{
			((MockUrlHelper)UrlHelper).ActionDelegate = actionCallBack;
		}
}

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs]: Both of these would certainly work. I would also add that `don't have to make MockUrlHelper public` isn't a design goal, if there's a benefit to making it public then we'll go that route.

Sometimes context is good.

`context.Url.OnAction(...)` vs `context.OnUrlAction(...)`

File: src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs (hunk 22 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: @rynowak I think this design has a problem when users call different overloaded Action methods in a controller. 1) It would make us to maintain a list of delegates for overloaded methods, 2) it makes user to write different callbacks for different overloaded delegates. 
Instead, I'd like to create a context class like
public class UrlHelperContext
{
    string Action {get;set;} string Controller {get; set;...}
}
And ..public Func<UrlHelperContext, string> OnUrlAction { get; set; }

IUrlHeper.RouteUrl method requires the same parameters, I think it makes APIs simpler after all. 

Let me know if you have any concerns on this.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelperContext.cs]: Would love to see you try this out and show an example. Give it a shot

File: test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs (hunk 13 above)
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: @rynowak 2 examples that you pointed out are really full-blown controllers. For now, I'm going to use Mock as much as possible for the all external components such as EF, UserManager, and so on. Plus, those tests are not going to be checked in the repository, but they are just to show how users codes look like..still I think it will takes some time..

Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Testing.Test/ControllerTestHelperTest.cs]: @rynowak I looked into two controllers, AccountController and AlubmsApiController. I don't think we are interested in all the actions of the controllers. Many of them doesn't even need this feature to test. 
 And, I think the following actions should be good enough to show how users codes look like:
1. VerifyCode for using ModelState.isValid and ModelState.AddModelError
2. Register for Url.Action, RedirectToAction, ViewBag.Link View, AddErrors
3. ExternalLoginConfirmation for User.Identity.IsAuthenticated and  ModelState.IsValid

Let me know if you have any other actions to try out.

**Comment by ianhong** [comment]
@yishaigalatzer and @rynowak 

I think this Design PR has all the feature to support MusicStore and KEpense except:

1. Context.User.Identity.GetUserId(),
2. Content.Request.Cookies

1. Contxt.User.Identity.GetUserId will be easy to support with this design
2. I haven't looked into Cookies yet. 

Thanks,

**Comment by ianhong** [comment]
Opened #1935 and #1934 for more specific work items, and closing this one.
